src/pages/index.js

```diff
@@ -15,80 +15,194 @@ import Section from "../components/Section";
-
 const editUserInfoPopup = new PopupWithForm(
   selectors.editProfileModal,
   handleProfileEditSubmit
 );
 
+const editAvatarPopup = new PopupWithForm(
+  selectors.editAvatarModal,
+  handleAvatarEditSubmit
+);
+
+const deleteCardPopup = new PopupConfirmDelete(selectors.confirmDeleteModal);
+
+const api = new Api({
+  baseUrl: "https://around-api.en.tripleten-services.com/v1",
+  headers: {
+    authorization: "2f5abbfd-693d-4973-87a8-2e8fc711af51",
+    "Content-Type": "application/json",
+  },
+});
+
 //Initialize all instances
-cardSection.renderItems();
 imagePreviewPopup.setEventListeners();
 newCardPopup.setEventListeners();
 editUserInfoPopup.setEventListeners();
+editAvatarPopup.setEventListeners();
+deleteCardPopup.setEventListeners();
+
+let cardSection;
+let userInfo;
+
+api
+  .initialPageLoad()
+  .then(([userData, cards]) => {
+    cardSection = new Section(
+      {
+        items: cards,
+        renderer: (item) => {
+          cardSection.addItem(createCard(item));
+        },
+      },
+      selectors.cardSection
+    );
+    cardSection.renderItems();
+
+    userInfo = new UserInfo(profileSelectors);
+    userInfo.setUserInfo(userData.name, userData.about);
+    userInfo.setUserAvatar(userData.avatar);
+  })
+  .catch((err) => {
+    console.error(`There is an Error: `, err); // log the error to the console
+  });
 
 //Specific Event Listeners
 addNewCardButton.addEventListener("click", () => {
   formValidators["place-form"].toggleButtonState();
   newCardPopup.open();
 });
 
+let userData;
+
 editProfileButton.addEventListener("click", () => {
-  const userInfo = editUserInfo.getUserInfo();
+  userData = userInfo.getUserInfo();
   editUserInfoPopup.setInputValues({
-    title: userInfo.name,
-    description: userInfo.job,
+    title: userData.name,
+    description: userData.job,
   });
   formValidators["profile-form"].resetValidation();
   editUserInfoPopup.open();
 });
 
+editAvatarButton.addEventListener("click", () => {
+  userData = userInfo.getUserInfo();
+  editAvatarPopup.setInputValues({
+    link: userData.src,
+  });
+  formValidators["avatar-form"].toggleButtonState();
+  editAvatarPopup.open();
+});
+
 // Functions
 function handleImageClick(link, name) {
   imagePreviewPopup.open(link, name);
 }
 
 function handleAddCardSubmit(cardData) {
-  cardSection.addItem(
-    createCard({ name: cardData.title, link: cardData.link })
-  );
+  newCardPopup.setButtonText(true);
+  api
+    .addNewCard(cardData.title, cardData.link)
+    .then((data) => {
+      cardSection.addItem(createCard(data));
+    })
+    .catch((err) => {
+      console.error(`There is an Error: `, err); // log the error to the console
+    })
+    .finally(() => {
+      newCardPopup.setButtonText(false);
+    });
+
   newCardPopup.close();
 }
 
 function handleProfileEditSubmit(cardData) {
-  editUserInfo.setUserInfo(cardData.title, cardData.description);
+  editUserInfoPopup.setButtonText(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```